app/src/main/java/io/gnosis/safe/ui/transaction/TransactionsViewModel.kt

```diff
@@ -59,24 +75,361 @@ class TransactionsViewModel
 
     private fun mapTransactions(
         transactions: Page<Transaction>,
-        safe: Solidity.Address
+        safeInfo: SafeInfo
     ): List<TransactionView> {
         return transactions.results.mapNotNull { transaction ->
-            when (transaction) {
-                is Transaction.Transfer -> TransactionView.Transfer(transaction, transaction.recipient == safe)
-                is Transaction.SettingsChange -> TransactionView.SettingsChange(transaction)
+            when {
+                transaction is Transfer && isHistoricTransfer(transaction) -> historicTransfer(transaction, safeInfo)
+                transaction is Transfer && isQueuedTransfer(transaction) -> queuedTransfer(transaction, safeInfo)
+                transaction is SettingsChange && isQueuedMastercopyChange(transaction) -> queuedMastercopyChange(transaction, safeInfo.threshold)
+                transaction is SettingsChange && isHistoricMastercopyChange(transaction) -> historicMastercopyChange(transaction)
+                transaction is SettingsChange && isQueuedSetFallbackHandler(transaction) -> queuedSetFallbackHandler(transaction, safeInfo.threshold)
+                transaction is SettingsChange && isHistoricSetFallbackHandler(transaction) -> historicSetFallbackHandler(transaction)
+                transaction is SettingsChange && isQueuedModuleChange(transaction) -> queuedModuleChange(transaction, safeInfo.threshold)
+                transaction is SettingsChange && isHistoricModuleChange(transaction) -> historicModuleChange(transaction)
+                transaction is SettingsChange && isQueuedSettingsChange(transaction) -> queuedSettingsChange(transaction, safeInfo.threshold)
+                transaction is SettingsChange && isHistoricSettingsChange(transaction) -> historicSettingsChange(transaction)
+                transaction is Custom && isQueuedCustomTransaction(transaction) -> queuedCustomTransaction(transaction, safeInfo.threshold)
+                transaction is Custom && isHistoricCustomTransaction(transaction) -> historicCustomTransaction(transaction)
                 else -> null
             }
         }
     }
 
+    private fun isHistoricSetFallbackHandler(settingsChange: SettingsChange): Boolean {
+        return settingsChange.isSetFallBackHandler() && settingsChange.isCompleted()
```

Comment:
maybe the method that is part of the `SettingsChange` class at the moment, should be inlined here.

Reply:
Done